service/dap/server_test.go

```diff
@@ -375,6 +388,70 @@ func expectStackFrames(t *testing.T, got *dap.StackTraceResponse,
 	}
 }
 
+// expectScope is a helper for verifying the values within a ScopesResponse.
+//     i - index of the scope within ScopesRespose.Body.Scopes array
+//     name - name of the scope
+//     varRef - reference to retrieve variables of this scope
+func expectScope(t *testing.T, got *dap.ScopesResponse, i int, name string, varRef int) {
+	t.Helper()
+	if len(got.Body.Scopes) <= i {
+		t.Errorf("\ngot  %d\nwant len(Scopes)>%d", len(got.Body.Scopes), i)
+	}
+	goti := got.Body.Scopes[i]
+	if goti.Name != name || goti.VariablesReference != varRef || goti.Expensive {
+		t.Errorf("\ngot  %#v\nwant Name=%q VariablesReference=%d Expensive=false", goti, name, varRef)
+	}
+}
+
+// expectVars is a helper for verifying the number of variables withing a VariablesResponse.
+//      parentName - name of the enclosing variable or scope
+//      numChildren - number of variables/fields/elements of this variable
+func expectVars(t *testing.T, got *dap.VariablesResponse, parentName string, numChildren int) {
+	t.Helper()
+	if len(got.Body.Variables) != numChildren {
+		t.Errorf("\ngot  len(%s)=%d\nwant %d", parentName, len(got.Body.Variables), numChildren)
+	}
+}
+
+// expectVar is a helper for verifying the values within a VariablesResponse.
+//     i - index of the scope within VariablesRespose.Body.Variables array
+//     name - name of the variable
+//     value - the value of the variable
+//     useExactMatch - true if value is to be compared to exactly, false if to be used as regex
+//     ref - reference to retrieve children of this variable
+func expectVar(t *testing.T, got *dap.VariablesResponse, i int, name, value string, useExactMatch bool, ref int) {
```

Comment:
I think you should support `i == -1`, where you scan `got` for the first variable with that name, and then use this everywhere the order of variables doesn't matter. I think as it is this set of tests is too fragile. For example if the definition of the chan struct changes a little bit (because a field gets inserted) the tests will fail. Same thing for the order of variables returned by `LocalVariables`.

Same thing for `ref`, allow `ref < 0` to match anything and return the `VariableReference` field of the variable that's been tested, so that tests don't have to depend on variable references being produced in a certain order.

Reply:
Yes, that's a very good idea. And it would definitely make it easier to update the tests. Will give this a try next and push in a separate commit.

Reply:
Do you mean a separate commit in this PR or a separate PR?

Reply:
Already pushed a commit in this PR last night.